govcd/entity_client.go

```diff
@@ -0,0 +1,225 @@
+package govcd
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+
+	"github.com/vmware/go-vcloud-director/v3/ccitypes"
+	"github.com/vmware/go-vcloud-director/v3/types/v56"
+	"github.com/vmware/go-vcloud-director/v3/util"
+)
+
+func (client *Client) GetEntityUrl(endpoint ...string) (*url.URL, error) {
+	path := fmt.Sprintf(ccitypes.KubernetesSubpath, client.VCDHREF.Scheme, client.VCDHREF.Host)
+	path = path + strings.Join(endpoint, "")
+
+	return url.ParseRequestURI(path)
+}
+
+func (client *Client) PostEntity(urlRef *url.URL, params url.Values, payload, outType interface{}, additionalHeader map[string]string) error {
+	// copy passed in URL ref so that it is not mutated
+	urlRefCopy := copyUrlRef(urlRef)
+
+	util.Logger.Printf("[TRACE] Posting %s item to endpoint %s with expected response of type %s",
+		reflect.TypeOf(payload), urlRefCopy.String(), reflect.TypeOf(outType))
+
+	if !client.IsTm() {
+		return fmt.Errorf("Client is not supported on this version")
+	}
+
+	resp, err := client.performEntityPostPut(http.MethodPost, urlRefCopy, params, payload, additionalHeader)
+	if err != nil {
+		return err
+	}
+
+	if err = decodeBody(types.BodyTypeJSON, resp, outType); err != nil {
+		return fmt.Errorf("error decoding JSON response after POST: %s", err)
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("error closing response body: %s", err)
+	}
+
+	return nil
+}
+
+func (client *Client) GetEntity(urlRef *url.URL, params url.Values, outType interface{}, additionalHeader map[string]string) error {
+	// copy passed in URL ref so that it is not mutated
+	urlRefCopy := copyUrlRef(urlRef)
+
+	util.Logger.Printf("[TRACE] Getting item from endpoint %s with expected response of type %s",
+		urlRefCopy.String(), reflect.TypeOf(outType))
+
+	if !client.IsTm() {
+		return fmt.Errorf("Client is not supported on this version")
+	}
+
+	req := client.newEntityRequest(params, http.MethodGet, urlRefCopy, nil, additionalHeader)
+	resp, err := client.Http.Do(req)
+	if err != nil {
+		return fmt.Errorf("error performing GET request to %s: %s", urlRefCopy.String(), err)
+	}
+
+	// HTTP 404: Not Found - include sentinel log message ErrorEntityNotFound so that errors can be
+	// checked with ContainsNotFound() and IsNotFound()
+	if resp.StatusCode == http.StatusNotFound {
+		err := ParseErr(types.BodyTypeJSON, resp, &ccitypes.ApiError{})
+		closeErr := resp.Body.Close()
+		return fmt.Errorf("%s: %s [body close error: %s]", ErrorEntityNotFound, err, closeErr)
+	}
+
+	// resp is ignored below because it is the same as above
+	_, err = checkRespWithErrType(types.BodyTypeJSON, resp, err, &ccitypes.ApiError{})
+
+	// Any other error occurred
+	if err != nil {
+		return fmt.Errorf("error in HTTP GET request for: %s", err)
+	}
+
+	if err = decodeBody(types.BodyTypeJSON, resp, outType); err != nil {
+		return fmt.Errorf("error decoding JSON response after GET: %s", err)
+	}
+
+	err = resp.Body.Close()
```

Comment:
Should we move this earlier with a defer? the above calls to `checkRespWithErrType` and `decodeBody` will also read the body, and I don't think we're closing it in case of an error.